File: viridian/algae/sources/coordinator.py
from asyncio import FIRST_COMPLETED, CancelledError, create_subprocess_shell, create_task, sleep, wait
from contextlib import suppress
from ipaddress import AddressValueError, IPv4Address
from os import getenv
from socket import AF_INET, SOCK_DGRAM, gaierror, gethostbyname, gethostname, inet_aton, socket
from subprocess import PIPE
from typing import Any, Dict, Optional

from colorama import Fore
from Crypto.Random import get_random_bytes
from Crypto.Random.random import randint
from grpclib.exceptions import GRPCError

from sources.generated import ControlHandshakeRequest, ControlHealthcheck, WhirlpoolAuthenticationRequest, WhirlpoolViridianStub
from sources.tunnel import Tunnel
from sources.utils import MAX_TAIL_LENGTH, SYMM_KEY_LENGTH, create_grpc_secure_channel, logger
from sources.viridian import Viridian

# Current algae distribution version.
VERSION = "0.0.3"

# Default tunnel interface name.
_DEFAULT_TUNNEL_NAME = "seatun"

# Default tunnel interface address.
_DEFAULT_TUNNEL_ADDRESS = "192.168.0.65"

# Default tunnel interface netmask.
_DEFAULT_TUNNEL_NETMASK = "255.255.255.0"

# Default tunnel interface seaside-viridian-algae code.
_DEFAULT_TUNNEL_SVA = 65

# Minimal time between two healthpings, in seconds.
_DEFAULT_HEALTHCHECK_MIN_TIME = 1

# Maximal time between two healthpings, in seconds.
_DEFAULT_HEALTHCHECK_MAX_TIME = 5

# Default gRPC maximal request timeout.
_DEFAULT_CONNECTION_TIMEOUT = 3.0


class Coordinator:
    """
    Viridian "coordinator" class: it is responsible for VPN connection management and lifecycle.
    It is capale of receiving connection token from viridian, initializing and supporting connection control.
    Externally, it connects to viridian node, sends and receives healthpings and terminates connection.
    Internally, it manages both "interface" and "client" viridian objects lifetime.
    In particular, during initialization it creates tunnel interface and opens client seaside port.
    They are closed and deleted in destructor or using a special function.
    They can also be stopped and resumed in runtime.
    """

    def __init__(self, payload: str, addr: str, ctrl_port: int):
        """
        Coordinator constructor.
        :param self: instance of Coordinator.
        :param payload: payload value string.
        :param addr: caerulean internal address.
        :param ctrl_port: caerulean gRPC control port.
        :param name: tunnel interface name.
        """
        try:
            self._address = str(IPv4Address(addr))
        except AddressValueError:
            self._address = gethostbyname(addr)

        self._viridian: Optional[Viridian] = None

        tunnel_name = getenv("SEASIDE_TUNNEL_NAME", _DEFAULT_TUNNEL_NAME)
        tunnel_address = IPv4Address(getenv("SEASIDE_TUNNEL_ADDRESS", _DEFAULT_TUNNEL_ADDRESS))
        tunnel_netmask = IPv4Address(getenv("SEASIDE_TUNNEL_NETMASK", _DEFAULT_TUNNEL_NETMASK))
        tunnel_sva = int(getenv("SEASIDE_TUNNEL_SVA", _DEFAULT_TUNNEL_SVA))
        self._tunnel = Tunnel(tunnel_name, tunnel_address, tunnel_netmask, tunnel_sva, IPv4Address(self._address))

        self._node_payload = payload
        self._ctrl_port = ctrl_port
        self._user_name = getenv("SEASIDE_USER_NAME", gethostname())
        self._min_hc_time = int(getenv("SEASIDE_MIN_HC_TIME", _DEFAULT_HEALTHCHECK_MIN_TIME))
        self._max_hc_time = int(getenv("SEASIDE_MAX_HC_TIME", _DEFAULT_HEALTHCHECK_MAX_TIME))
        self._max_timeout = float(getenv("SEASIDE_CONNECTION_TIMEOUT", _DEFAULT_CONNECTION_TIMEOUT))

        if self._min_hc_time < 1:
            raise ValueError("Minimal healthcheck time can't be less than 1 second!")
        if self._max_hc_time < 1:
            raise ValueError("Maximum healthcheck time can't be less than 1 second!")

        self._gate_socket = socket(AF_INET, SOCK_DGRAM)
        self._gate_socket.bind((self._tunnel.default_ip, 0))
        self._gate_socket.setblocking(False)

        authority = getenv("SEASIDE_ROOT_CERTIFICATE_AUTHORITY", None)
        self._channel = create_grpc_secure_channel(self._address, self._ctrl_port, authority)
        self._control = WhirlpoolViridianStub(self._channel)

        self._user_id: int
        self._session_token: bytes
        self._session_key: bytes

    async def _initialize_connection(self) -> None:
        """
        Open and start "interface" and "viridian" objects.
        Also receive connection token and connect to caerulean.
        Clean tunnel interface in case of any error.
        """
        try:
            if self._viridian is not None and self._viridian.operational:
                logger.info("Closing the seaside client...")
                self._viridian.close()
            logger.info("Receiving user token...")
            await self._receive_token()
            logger.info("Exchanging basic information...")
            await self._initialize_control()
            if not self._tunnel.operational:
                logger.info("Opening the tunnel...")
                self._tunnel.up()
            if self._viridian is not None:
                logger.info("Opening the seaside client...")
                self._viridian.open()
        except BaseException:
            self._clean_tunnel()
            raise

    async def _run_vpn_loop(self) -> None:
        """
        Create VPN connection.
        Receive viridian connection token, initialize and manage control.
        Upon receiving an error message, client is re-initialized, token is received once again and control is re-initialized.
        NB! This method should be run while VPN is active.
        """
        logger.info("Running VPN loop...")
        while self._tunnel.operational:
            try:
                logger.info("Sending healthcheck request...")
                await self._perform_control()
            except GRPCError:
                logger.info("Control error occurs, trying to reconnect!")
                logger.info("Re-initializing connection...")
                await self._initialize_connection()
            except BaseException as exc:
                logger.debug(f"Interrupting connection to caerulean {self._address}:{self._ctrl_port}...")
                await self.interrupt()
                raise exc

    async def _run_vpn_command(self, cmd: str) -> None:
        """
        Execute command asynchronously, wait for the result, print the result code.
        In case command produces STDOUT or STDERR output, it will also be printed.
        :param cmd: command that will be run.
        """
        proc = await create_subprocess_shell(cmd, stdout=PIPE, stderr=PIPE)
        stdout, stderr = await proc.communicate()
        print(f"The command exited with: {proc.returncode}")
        if len(stdout) > 0:
            print(f"STDOUT: {stdout.decode()}")
        if len(stderr) > 0:
            print(f"STDERR: {stderr.decode()}")

    async def start(self, cmd: Optional[str]) -> None:
        """
        Start VPN.
        Will open a VPN connection and launch the command that should be executed while VPN is active.
        If the command is specified, VPN connection will be terminated once it is finished (no matter what was the result).
        Also, a DNS probe will be made once connection is opened to ensure DNS servers are still accessible.
        :param cmd: command that will be run while VPN is active.
        """
        logger.info("Initializing connection...")
        await self._initialize_connection()

        try:
            gethostbyname("example.com")
        except gaierror:
            logger.warning("WARNING! DNS probe failed! It is very likely that you have local DNS servers configured only!")

        task_set = set()
        task_set.add(create_task(self._run_vpn_loop()))
        if cmd is not None:
            task_set.add(create_task(self._run_vpn_command(cmd)))

        _, pending = await wait(task_set, return_when=FIRST_COMPLETED)
        for p in pending:
            p.cancel()
            with suppress(CancelledError):
                await p

    def _grpc_metadata(self) -> Dict[str, Any]:
        """
        Generate gRPC tail metadata.
        It consists of random number of random bytes.
        :return: gRPC metadata dictionary.
        """
        tail_metadata = ("seaside-tail-bin", get_random_bytes(randint(1, MAX_TAIL_LENGTH)))
        return {"timeout": self._max_timeout, "metadata": (tail_metadata,)}

    async def _receive_token(self) -> None:
        """
        Receive viridian connection token.
        Alongside with token, receive caerulean seaside and control port numbers.
        Also initialize the session obfuscator.
        """
        self._session_key = get_random_bytes(SYMM_KEY_LENGTH)
        logger.debug(f"Symmetric session cipher initialized ({self._user_name}): {self._session_key!r}!")
        request = WhirlpoolAuthenticationRequest(self._user_name, self._session_key, self._node_payload)

        logger.debug("Requesting whirlpool token...")
        response = await self._control.authenticate(request, **self._grpc_metadata())
        self._session_token = response.token
        logger.debug(f"Symmetric session token received: {self._session_token!r}!")

        if response.max_next_in < self._min_hc_time:
            self._min_hc_time = response.max_next_in
            logger.debug(f"Minimum healthcheck delay updated to: {self._min_hc_time}!")
        if response.max_next_in < self._max_hc_time:
            self._max_hc_time = response.max_next_in
            logger.debug(f"Maximum healthcheck delay updated to: {self._max_hc_time}!")


        if response.max_next_in < self._min_hc_time:
            self._min_hc_time = response.max_next_in
            logger.debug(f"Minimum healthcheck delay updated to: {self._min_hc_time}!")
        if response.max_next_in < self._max_hc_time:
            self._max_hc_time = response.max_next_in
            logger.debug(f"Maximum healthcheck delay updated to: {self._max_hc_time}!")

    async def _initialize_control(self) -> None:
        """
        Handshake with VPN node and initialize connection control.
        Initialize "viridian" object.
        Only proceed if valid user ID and successful control response status is received.
        """
        logger.debug(f"Making handshake caerulean {self._address}:{self._ctrl_port}...")
        request = ControlHandshakeRequest(self._session_token, VERSION, self._node_payload, inet_aton(self._tunnel.default_ip), self._gate_socket.getsockname()[1])
        response = await self._control.handshake(request, **self._grpc_metadata())

        if response.user_id is None:
            raise RuntimeError("User ID is None in control server response!")
        else:
            logger.info(f"User ID assigned: {Fore.BLUE}{response.user_id}{Fore.RESET}")
            self._user_id = response.user_id

        self._viridian = Viridian(self._gate_socket, self._tunnel.descriptor, self._address, self._session_key, self._user_id)
        logger.info(f"Handshake with caerulean {self._address}:{self._ctrl_port} completed successfully!")

    def _clean_tunnel(self) -> None:
        """
        Close both "interface" and "client" objects if they are still running.
        Also close the seaside socket and delete "interface".
        """
        logger.info("Terminating whirlpool connection...")
        if self._viridian is not None and self._viridian.operational:
            logger.info("Closing the seaside client...")
            self._viridian.close()
        if self._tunnel.operational:
            logger.info("Closing the tunnel...")
            self._tunnel.down()
        logger.info("Closing the seaside socket...")
        self._gate_socket.close()

    async def _perform_control(self) -> None:
        """
        Exchange healthping messages and sleep until the next healthping message is ready.
        """
        next_in = randint(self._min_hc_time, self._max_hc_time)
        request = ControlHealthcheck(user_id=self._user_id, next_in=next_in)
        await self._control.healthcheck(request, **self._grpc_metadata())
        logger.info(f"Healthcheck performed, sleeping for {next_in} seconds!!")
        await sleep(next_in)

    async def interrupt(self) -> None:
        """
        Interrupt VPN connection gracefully.
        Includes not only tunnel closing ("interface", "viridian" and seaside socket), but also sending termination request to caerulean.
        Finally, removes tunnel interface.
        """
        logger.debug(f"Interrupting connection to caerulean {self._address}:{self._ctrl_port}...")

        self._channel.close()
        self._clean_tunnel()
        logger.warning("Whirlpool connection terminated!")

        self._tunnel.delete()
        logger.warning("Local viridian interface removed!!")

File: viridian/reef/src/lib/coordinator.rs
use std::error::Error;
use std::net::Ipv4Addr;
use std::time::Duration;
use std::fs::read_to_string;

use rand::prelude::{thread_rng, RngCore, Rng};
use rand::rngs::ThreadRng;
use tokio::net::UdpSocket;
use tokio::runtime::Handle;
use tokio::task::block_in_place;
use tokio::time::sleep;
use tonic::metadata::MetadataValue;
use tonic::transport::{Certificate, Channel, ClientTlsConfig};
use tonic::Request;

use generated::whirlpool_viridian_client::WhirlpoolViridianClient;
use generated::{WhirlpoolAuthenticationRequest, WhirlpoolAuthenticationResponse, ControlHandshakeRequest, ControlHandshakeResponse, ControlHealthcheck, ControlException, ControlExceptionStatus};

use super::tunnel::Tunnel;
use super::viridian::Viridian;
use super::VERSION;

mod generated {
    tonic::include_proto!("generated");
}

const GRPC_PROTOCOL: &str = "https";
const MAX_TAIL_LENGTH: usize = 64;
const SYMM_KEY_LENGTH: usize = 32;
const NONE_USER_ID: u16 = 0;


pub struct Coordinator {
    tunnel: Tunnel,
    viridian: Option<Viridian>,
    socket: UdpSocket,
    client: WhirlpoolViridianClient<Channel>,

    node_payload: String,
    user_name: String,
    user_id: u16,
    min_hc_time: u16,
    max_hc_time: u16,

    session_token: Option<Vec<u8>>,
    session_key: Option<Vec<u8>>,
    randomizer: ThreadRng
}


impl Coordinator {
    pub async fn new(address: Ipv4Addr, ctrl_port: u16, payload: &str, user_name: &str, min_hc_time: u16, max_hc_time: u16, max_timeout: f32, tunnel_name: &str, ca: Option<&str>) -> Result<Coordinator, Box<dyn Error>> {
        let viridian_host = format!("{GRPC_PROTOCOL}://{address}:{ctrl_port}");

        if min_hc_time < 1 {
            return Err(Box::from("Minimum healthcheck time shouldn't be less than 1 second!"));
        }
        if max_hc_time < 1 {
            return Err(Box::from("Maximum healthcheck time shouldn't be less than 1 second!"));
        }

        let tunnel = Tunnel::new(tunnel_name, address).await?;
        let socket = UdpSocket::bind((tunnel.default_interface().0, 0)).await?;

        let tls = match ca {
            Some(certificate) => ClientTlsConfig::new().ca_certificate(Certificate::from_pem(read_to_string(certificate)?.as_bytes())),
            None => ClientTlsConfig::new().with_webpki_roots()
        };
        let caerulean_max_timeout = Duration::from_secs_f32(max_timeout);
        let channel = Channel::from_shared(viridian_host)?.timeout(caerulean_max_timeout).connect_timeout(caerulean_max_timeout).tls_config(tls)?.connect().await?;
        let client = WhirlpoolViridianClient::new(channel);

        Ok(Coordinator {
            tunnel,
            viridian: None,
            socket,
            client,
            node_payload: payload.to_string(),
            user_name: user_name.to_string(),
            user_id: NONE_USER_ID,
            min_hc_time,
            max_hc_time,
            session_token: None,
            session_key: None,
            randomizer: thread_rng()
        })
    }

    async fn initialize_connection(&mut self) -> Result<u16, Box<dyn Error>> {
        if let Some(viridian) = &self.viridian {
            if viridian.is_operational() {
                viridian.close();
            }
        }

        let recevive_token_res = self.receive_token().await;
        if let Err(res) = recevive_token_res {
            println!("log receive token error status: {res}");
            return Err(Box::from("Token receiving failed"));
        }

        let initialize_control_res = self.initialize_control().await;
        if let Err(ref res) = initialize_control_res {
            println!("log initialize control error status: {res}");
            return Err(Box::from("Control initialization failed"));
        }

        if let Some(viridian) = &self.viridian {
            viridian.open();
        }
        Ok(initialize_control_res.unwrap())
    }

    pub async fn start(&mut self) -> Result<(), Box<dyn Error>> {
        self.user_id = self.initialize_connection().await?;
        loop {
            let control = self.perform_control(self.user_id).await;
            if let Err(ctrl) = control {
                println!("log error status: {ctrl}");
                self.user_id = self.initialize_connection().await?;
            }
        }
    }

    fn make_grpc_request<T>(&mut self, message: T) -> Request<T> {
        let mut tail = Vec::with_capacity(self.randomizer.gen_range(1..MAX_TAIL_LENGTH));
        self.randomizer.fill_bytes(&mut tail);

        let mut request = Request::new(message);
        request.metadata_mut().append_bin("seaside-tail-bin", MetadataValue::from_bytes(&tail));
        request
    }

    async fn receive_token(&mut self) -> Result<(), tonic::Status> {
        let mut session_key = vec![0; SYMM_KEY_LENGTH];
        self.randomizer.fill_bytes(&mut session_key);

        let message = WhirlpoolAuthenticationRequest {uid: self.user_name.clone(), session: session_key.clone(), payload: self.node_payload.clone()};
        let request = self.make_grpc_request(message);
        let response = self.client.authenticate(request).await;

        match response {
            Ok(resp) => {
                self.session_key = Some(session_key);
                self.session_token = Some(resp.get_ref().token.clone());
                Ok(())
            },
            Err(resp) => Err(resp)
        }
    }

    async fn initialize_control(&mut self) -> Result<u16, tonic::Status> {
        let message = ControlHandshakeRequest {
            token: self.session_token.clone().unwrap(),
            version: VERSION.to_string(),
            payload: Some(self.node_payload.clone()),
            address: self.tunnel.default_interface().0.octets().to_vec(),
            port: i32::from(self.socket.local_addr().ok().unwrap().port())
        };
        let request = self.make_grpc_request(message);
        let response = self.client.handshake(request).await;

        self.viridian = Some(Viridian::new());
        match response {
            Ok(resp) => Ok(resp.get_ref().user_id as u16),
            Err(resp) => Err(resp)
        }
    }

    async fn perform_control(&mut self, user_id: u16) -> Result<tonic::Response<()>, tonic::Status> {
        let next_in = self.randomizer.gen_range(self.min_hc_time..self.max_hc_time);
        let message = ControlHealthcheck {user_id: i32::from(user_id), next_in: i32::from(next_in)};
        let request = self.make_grpc_request(message);
        let response = self.client.healthcheck(request).await;
        sleep(Duration::from_secs(u64::from(next_in))).await;
        response
    }

    async fn interrupt(&mut self, exception: Option<String>) -> Result<(), Box<dyn Error>> {
        let message = ControlException {
            status: i32::from(ControlExceptionStatus::Termination),
            user_id: i32::from(self.user_id),
            message: exception
        };
        let request = self.make_grpc_request(message);
        let response = self.client.exception(request).await;
        match response {
            Ok(_) => Ok(()),
            Err(resp) => Err(Box::from(resp))
        }
    }
}

impl Drop for Coordinator {
    fn drop(&mut self) -> () {
        if self.user_id != NONE_USER_ID {
            block_in_place(move || {
                Handle::current().block_on(async move {
                    // TODO: log message!
                    self.interrupt(None).await.expect("...");
                });
            });
        }
    }
}
